external/mlir-hal/lib/Dialect/MHAL/Transforms/Prefill.cpp

```diff
@@ -0,0 +1,140 @@
+
+  Value createConstantOp(OpBuilder &builder, Type elementType,
+                         uint32_t initValue) {
+    Value constantOp;
+    auto loc = builder.getUnknownLoc();
+    if (llvm::isa<FloatType>(elementType)) {
+      auto semantics = static_cast<APFloat::Semantics>(-1);
+      if (elementType.isF32()) {
+        semantics = APFloat::S_IEEEsingle;
+      } else if (elementType.isF16()) {
+        semantics = APFloat::S_IEEEhalf;
+      } else if (elementType.isBF16()) {
+        semantics = APFloat::S_BFloat;
+      } else if (elementType.isFloat8E4M3FNUZ()) {
+        semantics = APFloat::S_Float8E4M3FNUZ;
+      } else if (elementType.isFloat8E5M2FNUZ()) {
+        semantics = APFloat::S_Float8E5M2FNUZ;
+      } else {
+        llvm_unreachable("Unexpected float semantics");
+      }
+
+      APFloat apValue(*(reinterpret_cast<float *>(&initValue)));
+      bool lostInfo = false;
+      apValue.convert(APFloat::EnumToSemantics(semantics),
+                      APFloat::rmNearestTiesToEven, &lostInfo);
+
+      auto constValue = builder.getFloatAttr(elementType, apValue);
+      constantOp =
+          builder.create<arith::ConstantOp>(loc, elementType, constValue);
+    } else {
+      APInt apValue(elementType.getIntOrFloatBitWidth(), initValue, true);
+      auto constValue = builder.getIntegerAttr(elementType, apValue);
+      constantOp =
+          builder.create<arith::ConstantOp>(loc, elementType, constValue);
+    }
+    return constantOp;
+  }
+
+  void insertPrefillOps(OpBuilder &builder, gpu::LaunchFuncOp &launchOp) {
+    auto func = llvm::cast<func::FuncOp>(launchOp->getParentOp());
+    auto module = llvm::cast<ModuleOp>(func->getParentOp());
+
+    auto kernel = launchOp.getKernel();
+    auto *callee = module.lookupSymbol(kernel);
+    assert(callee != nullptr && "expect to find the function defenition");
+
+    auto llvmFunc = llvm::cast<LLVM::LLVMFuncOp>(callee);
+    auto gpuModule = llvm::cast<gpu::GPUModuleOp>(llvmFunc->getParentOp());
+    assert(gpuModule != nullptr &&
+           "expect LLVMFuncOp to be inside GPUModuleOp");
+
+    SmallVector<mhal::PrefillAttr, 4> prefillAttrs;
+    if (auto moduleAttr = gpuModule->getAttr(llvmFunc.getSymName())) {
+      if (auto arrayAttr = llvm::dyn_cast<ArrayAttr>(moduleAttr)) {
+        for (auto attr : arrayAttr) {
+          if (auto prefillAttr = llvm::dyn_cast<mhal::PrefillAttr>(attr)) {
+            prefillAttrs.push_back(prefillAttr);
+          }
+        }
+      }
+    }
+
+    auto loc = builder.getUnknownLoc();
+    auto kernelOperands = launchOp.getKernelOperands();
+    for (auto attr : prefillAttrs) {
+      auto argIdx = attr.getArgIndex();
+      assert(argIdx < kernelOperands.size() &&
+             "provided arg index is out of bounds");
+
+      auto arg = kernelOperands[argIdx];
+      auto type = arg.getType().cast<MemRefType>();
+      auto elementType = type.getElementType();
+
+      builder.setInsertionPointToStart(arg.getParentBlock());
```

Comment:
I don't think that's right. I'd set the insertion point before the launch?

Also, async tokens?

Reply:
Hi @krzysz00,

there are two places where I user `setInsertionPoint`, namely: 1) At the beginning of the parent block to insert a constant provided by the `PrefillAttr` and 2) to insert `memcpy` before a kernel launch. 

You highlighted the first case - i.e., inserting a constant to the beginning of the parent block.

The whole code looks like the following

```cpp
    builder.setInsertionPointToStart(arg.getParentBlock());
    auto initConstant = builder.create<arith::ConstantOp>(loc, elementType,
                                                          attr.getInitValue());
    builder.setInsertionPoint(launchOp->getBlock(),
                              --Block::iterator(launchOp));
    builder.create<gpu::MemsetOp>(loc, mlir::Type{}, mlir::ValueRange{}, arg,
                                  initConstant);
```

But, I agree one can just use one `setInsertionPoint`. For example, 

```cpp
    builder.setInsertionPoint(launchOp->getBlock(),
                              --Block::iterator(launchOp));
    auto initConstant = builder.create<arith::ConstantOp>(loc, elementType,
                                                          attr.getInitValue());
    builder.create<gpu::MemsetOp>(loc, mlir::Type{}, mlir::ValueRange{}, arg,
                                  initConstant);
```

Both solution will work. However, the second one results in less code. Ok, I will just update my implementation according to the second case which I wrote above.